engine/CSVParser.cc

```diff
@@ -447,26 +461,85 @@ namespace minsky
-          throw runtime_error("memory threshold exeeded");
-        // stash the data into vv tensorInit field
-        v.tensorInit.data.clear();
-        v.tensorInit.data.resize(v.numElements(), spec.missingValue);
-        auto dims=v.tensorInit.dims=v.dims();    
-        for (auto& i: tmpData)
-          {
-            size_t idx=0;
-            assert (dims.size()==i.first.size());
-            assert(dimLabels.size()==dims.size());
-            for (int j=dims.size()-1; j>=0; --j)
-              {
-                assert(dimLabels[j].count(i.first[j]));
-                idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
-              }
-            v.tensorInit.data[idx]=i.second;
-          }
+                  
+        v.sparsityRatio =  static_cast<double>(1.0-static_cast<double>(tmpData.size())/v.numDenseElements()); 
+		
+		// for feature 47
+		if (v.sparsityRatio <= 0.5) 
+	    {
+          v.setXVector(xVector);          
+          if (!cminsky().checkMemAllocation(v.numDenseElements()*sizeof(double)))
+            throw runtime_error("memory threshold exceeded");
+		  v.index.clear();	            
+          // stash the data into vv tensorInit field
+          v.tensorInit.data.clear();
+          v.tensorInit.data.resize(v.numDenseElements(), spec.missingValue);  
+          auto dims=v.tensorInit.dims=v.dims();    
+          for (auto& i: tmpData)
+            {
+              size_t idx=0;
+              assert (dims.size()==i.first.size());
+              assert(dimLabels.size()==dims.size());
+              for (int j=dims.size()-1; j>=0; --j)
+                {
+                  assert(dimLabels[j].count(i.first[j]));
+                  idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
+                }
+              v.tensorInit.data[idx]=i.second;  
+            }
+        }    
+        else 
+	    {	
+		  v.numSparseElements=tmpData.size();	
+          v.setXVector(xVector);          
+          if (!cminsky().checkMemAllocation(v.numSparseElements*sizeof(double)))
+            throw runtime_error("memory threshold exceeded");		  
+		  v.index.clear();			
+		  v.tensorInit.data.clear();	
+		  v.tensorInit.data.resize(v.numSparseElements, spec.missingValue); 	
+		  vector<double> tempData;
+          auto dims=v.tensorInit.dims=v.dims();    
+          for (auto& i: tmpData)
+            {
+              size_t idx=0;
+              assert (dims.size()==i.first.size());
+              assert(dimLabels.size()==dims.size());
+              for (int j=dims.size()-1; j>=0; --j)
+                {
+                  assert(dimLabels[j].count(i.first[j]));
+                  idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
+                }
+               if (!isnan(i.second)) {
+                 v.index.push_back(idx);
+                 tempData.push_back(i.second);
+               }                   
+            }
+           
+            // for feature 47           
+            for (auto& i: tmpAll)
```

Comment:
What is the purpose of this loop?

Reply:
I thought I had to run through all the serialized hypercube indices (including those of missing data) to pick out the onse corresponding to actual data. Of course this is not necessary because the loop over tmpData already does this. I've removed it.